model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,656 @@
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import static org.keycloak.models.UserSessionModel.CORRESPONDING_SESSION_ID;
+import static org.keycloak.models.UserSessionModel.SessionPersistenceState.TRANSIENT;
+import static org.keycloak.utils.StreamsUtil.paginatedStream;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public class MapUserSessionProvider implements UserSessionProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapUserSessionProvider.class);
+    private final KeycloakSession session;
+    protected final MapKeycloakTransaction<UUID, MapUserSessionEntity, UserSessionModel> userSessionTx;
+    protected final MapKeycloakTransaction<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionTx;
+    private final MapStorage<UUID, MapUserSessionEntity, UserSessionModel> userSessionStore;
+    private final MapStorage<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionStore;
+
+    /**
+     * Storage for transient user sessions which lifespan is limited to one request.
+     */
+    private final Map<UUID, MapUserSessionEntity> transientUserSessions = new HashMap<>();
+
+    public MapUserSessionProvider(KeycloakSession session, MapStorage<UUID, MapUserSessionEntity, UserSessionModel> userSessionStore,
+                                  MapStorage<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionStore) {
+        this.session = session;
+        this.userSessionStore = userSessionStore;
+        this.clientSessionStore = clientSessionStore;
+        userSessionTx = userSessionStore.createTransaction(session);
+        clientSessionTx = clientSessionStore.createTransaction(session);
+
+        session.getTransactionManager().enlistAfterCompletion(userSessionTx);
+        session.getTransactionManager().enlistAfterCompletion(clientSessionTx);
+    }
+
+    private Function<MapUserSessionEntity, UserSessionModel> userEntityToAdapterFunc(RealmModel realm) {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return (origEntity) -> new MapUserSessionAdapter(session, realm,
+                Objects.equals(origEntity.getPersistenceState(), TRANSIENT) ? origEntity : registerEntityForChanges(origEntity)) {
+
+            @Override
+            public void removeAuthenticatedClientSessions(Collection<String> removedClientUUIDS) {
+                removedClientUUIDS.forEach(clientId -> {
+                    clientSessionTx.delete(origEntity.getAuthenticatedClientSessions().get(clientId));
+                    entity.removeAuthenticatedClientSession(clientId);
+                });
```

Comment:
Ensure that this is using bulk delete, e.g. 
```suggestion
                ModelCriteriaBuilder<AuthenticatedClientSessionModel> mcb = clientSessionStore.createCriteriaBuilder()
                  .compare(AuthenticatedClientSessionModel.SearchableFields.ID, Operator.IN,
                    removedClientUUIDS.stream().map(origEntity.getAuthenticatedClientSessions()::get).filter(Objects::nonNull));
                clientSessionTx.delete(UUID.randomUUID(), mcb);

                removedClientUUIDS.forEach(entity::removeAuthenticatedClientSession);

```